### PR TITLE
ARROW-5560: [C++][Plasma] Cannot create Plasma object after OutOfMemory error

### PR DESCRIPTION
**Pull request by richardliaw**
If the client tries to call `CreateObject` and there is not enough memory left in the object store to create it, an `OutOfMemory` error will be returned. 

However, the plasma store also creates an entry for the object, even though it failed to be created. This means that later on, if the client tries to create the object again, it will receive an error that the object already exists. Also, if the client tries to get the object, it will hang because the entry appears to be unsealed.

We fix this by only creating the object entry if the `CreateObject` operation succeeds.

### CODE REVIEW

**Comment by richardliaw** [comment]
cc @stephanie-wang

File: cpp/src/plasma/store.cc [hunk 1]
```diff
@@ -259,6 +261,10 @@ PlasmaError PlasmaStore::CreateObject(const ObjectID& object_id, int64_t data_si
   entry->create_time = std::time(nullptr);
   entry->construct_duration = -1;
 
+#ifdef PLASMA_CUDA
+  entry->ipc_handle = ipc_handle;
```
Comment (stephanie-wang) [inline cpp/src/plasma/store.cc]:
```suggestion
  entry->ipc_handle = result->ipc_handle;
```
Looks like `ipc_handle` wouldn't have been defined here.

**Comment by fsaintjacques** [comment]
The diff looks sound to me, were you able to reproduce with a unit test?

**Comment by richardliaw** [comment]
I have a unit test script that works on the Ray side and does not pass without this patch - do I need to push a unit test on the Arrow side?

**Comment by pcmoritz** [comment]
Do you want to come up with a small regression test that does the followin:

- Start object store with small amount of memory
- Create an object that is slightly larger than this memory, with a known object id
- Have it fail
- Try to recreate an object with the same id

That would be helpful! You can look at `test_plasma.py` for examples on how these tests look like (and the regression test should also go in there).

**Comment by richardliaw** [comment]
@pcmoritz can you take a look at the test introduced? I've verified that this fails on master and works on this branch.

**Comment by pcmoritz** [comment]
Yeah, I think it looks good and can be merged when the tests pass!

**Comment by pcmoritz** [comment]
There is a linting error (the other error is from arrow.flight and is unrelated I think).

**Comment by pcmoritz** [comment]
Merged, thanks for the contribution :)

**Pull request closed without merging**